File: ruby/nxs.rb
# frozen_string_literal: true
# NXS Reader — .nxb parser (Ruby 3.x, stdlib only).
#
# Implements Nexus Standard v1.0 binary wire format.
#
# Usage:
#   buf = File.binread("data.nxb")
#   reader = Nxs::Reader.new(buf)
#   reader.record_count          # => Integer
#   reader.keys                  # => Array<String>
#   obj = reader.record(42)      # => Nxs::Object
#   obj.get_str("username")      # => String | nil
#   obj.get_i64("id")            # => Integer | nil
#   obj.get_f64("score")         # => Float | nil
#   obj.get_bool("active")       # => true/false | nil
#   reader.sum_f64("score")      # => Float
#   reader.min_f64("score")      # => Float | nil
#   reader.max_f64("score")      # => Float | nil
#   reader.sum_i64("id")         # => Integer

module Nxs
  MAGIC_FILE   = 0x4E585342  # NXSB
  MAGIC_OBJ    = 0x4E58534F  # NXSO
  MAGIC_FOOTER = 0x2153584E  # NXS!
  FLAG_SCHEMA  = 0x0002

  class NxsError < StandardError
    attr_reader :code
    def initialize(code, msg)
      super("#{code}: #{msg}")
      @code = code
    end
  end

  # ── Reader ──────────────────────────────────────────────────────────────────

  class Reader
    attr_reader :keys, :record_count

    def initialize(bytes)
      @data = bytes.b   # force binary encoding
      sz = @data.bytesize
      raise NxsError.new("ERR_OUT_OF_BOUNDS", "file too small") if sz < 32

      magic = @data.unpack1("L<")
      raise NxsError.new("ERR_BAD_MAGIC", "expected NXSB, got 0x#{magic.to_s(16)}") if magic != MAGIC_FILE

      footer = @data.unpack1("@#{sz - 4}L<")
      raise NxsError.new("ERR_BAD_MAGIC", "footer magic mismatch") if footer != MAGIC_FOOTER

      # Preamble: Version(2) + Flags(2) + DictHash(8) + TailPtr(8) + Reserved(8)
      @flags    = @data.unpack1("@6 S<")
      @tail_ptr = @data.unpack1("@16 Q<")

      # Schema (when Flags bit 1 set)
      @keys       = []
      @key_sigils = []
      @key_index  = {}
      read_schema(32) if @flags & FLAG_SCHEMA != 0

      # Tail-index: u32 EntryCount followed by records
      @record_count = @data.unpack1("@#{@tail_ptr}L<")
      @tail_start   = @tail_ptr + 4
    end

    # O(1) record lookup — reads one 10-byte tail-index entry.
    def record(i)
      unless i >= 0 && i < @record_count
        raise NxsError.new("ERR_OUT_OF_BOUNDS", "record #{i} out of [0, #{@record_count})")
      end
      # Each tail-index entry: u16 KeyID + u64 AbsoluteOffset = 10 bytes
      abs_offset = @data.unpack1("@#{@tail_start + i * 10 + 2}Q<")
      Object.new(self, abs_offset)
    end

    # Tight allocation-free sum loop.
    def sum_f64(key)
      slot = @key_index[key]
      raise NxsError.new("ERR_OUT_OF_BOUNDS", "key '#{key}' not in schema") unless slot
      data = @data
      tail = @tail_start
      n    = @record_count
      sum  = 0.0
      i = 0
      while i < n
        abs = data.unpack1("@#{tail + i * 10 + 2}Q<")
        off = _scan_offset(data, abs, slot)
        sum += data.unpack1("@#{off}E") if off
        i += 1
      end
      sum
    end

    def min_f64(key)
      slot = @key_index[key]
      raise NxsError.new("ERR_OUT_OF_BOUNDS", "key '#{key}' not in schema") unless slot
      data = @data
      tail = @tail_start
      n    = @record_count
      min  = nil
      i = 0
      while i < n
        abs = data.unpack1("@#{tail + i * 10 + 2}Q<")
        off = _scan_offset(data, abs, slot)
        if off
          v   = data.unpack1("@#{off}E")
          min = v if min.nil? || v < min
        end
        i += 1
      end
      min
    end

    def max_f64(key)
      slot = @key_index[key]
      raise NxsError.new("ERR_OUT_OF_BOUNDS", "key '#{key}' not in schema") unless slot
      data = @data
      tail = @tail_start
      n    = @record_count
      max  = nil
      i = 0
      while i < n
        abs = data.unpack1("@#{tail + i * 10 + 2}Q<")
        off = _scan_offset(data, abs, slot)
        if off
          v   = data.unpack1("@#{off}E")
          max = v if max.nil? || v > max
        end
        i += 1
      end
      max
    end

    def sum_i64(key)
      slot = @key_index[key]
      raise NxsError.new("ERR_OUT_OF_BOUNDS", "key '#{key}' not in schema") unless slot
      data = @data
      tail = @tail_start
      n    = @record_count
      sum  = 0
      i = 0
      while i < n
        abs = data.unpack1("@#{tail + i * 10 + 2}Q<")
        off = _scan_offset(data, abs, slot)
        sum += data.unpack1("@#{off}q<") if off
        i += 1
      end
      sum
    end

    # Expose internals for Object
    attr_reader :data, :key_index

    # Walk the LEB128 bitmask from obj_offset+8, count set bits before `slot`,
    # and return the absolute byte offset of the field value (or nil if absent).
    # Used by both bulk reducers and NxsObject.
    def _scan_offset(data, obj_offset, slot)
      p = obj_offset + 8   # skip Magic(4) + Length(4)
      cur   = 0
      t_idx = 0

      loop do
        b    = data.getbyte(p)
        p   += 1
        bits = b & 0x7F
        7.times do |i|
          if cur == slot
            # field absent if bit is 0
            return nil if (bits >> i) & 1 == 0
            # p already past this bitmask byte; drain remaining continuation bytes
            while (b & 0x80) != 0
              b = data.getbyte(p)
              p += 1
            end
            # p now points to the offset table
            rel = data.unpack1("@#{p + t_idx * 2}S<")
            return obj_offset + rel
          end
          t_idx += 1 if (bits >> i) & 1 == 1
          cur += 1
        end
        # If all 7 bits processed and continuation bit clear, field is absent
        return nil if (b & 0x80) == 0
      end
    end

    private

    def read_schema(offset)
      key_count = @data.unpack1("@#{offset}S<")
      offset += 2

      @key_sigils = @data[offset, key_count].bytes
      offset += key_count

      # Null-terminated UTF-8 strings in StringPool
      pool = @data[offset..]
      pos  = 0
      key_count.times do |i|
        term = pool.index("\x00", pos)
        @keys << pool[pos...term].force_encoding("UTF-8")
        @key_index[@keys.last] = i
        pos = term + 1
      end
    end
  end

  # ── Object ───────────────────────────────────────────────────────────────────

  class Object
    def initialize(reader, offset)
      @reader = reader
      @offset = offset
      @parsed = false
    end

    def get_str(key)
      off = field_offset(key)
      return nil unless off
      len = @reader.data.unpack1("@#{off}L<")
      @reader.data[off + 4, len].force_encoding("UTF-8")
    end

    def get_i64(key)
      off = field_offset(key)
      return nil unless off
      @reader.data.unpack1("@#{off}q<")
    end

    def get_f64(key)
      off = field_offset(key)
      return nil unless off
      @reader.data.unpack1("@#{off}E")
    end

    def get_bool(key)
      off = field_offset(key)
      return nil unless off
      @reader.data.getbyte(off) != 0
    end

    private

    # Parse the object header (lazy — only on first field access).
    def parse_header
      return if @parsed
      p = @offset

      magic = @reader.data.unpack1("@#{p}L<")
      raise NxsError.new("ERR_BAD_MAGIC", "expected NXSO at #{p}") if magic != MAGIC_OBJ
      p += 8  # skip Magic(4) + Length(4)

      bitmask = []
      loop do
        b = @reader.data.getbyte(p)
        p += 1
        bitmask << (b & 0x7F)
        break if (b & 0x80) == 0
      end

      @bitmask          = bitmask
      @offset_tbl_start = p
      @parsed           = true
    end

    # Return the absolute byte offset of the field for `key`, or nil.
    def field_offset(key)
      slot = @reader.key_index[key]
      return nil unless slot

      # Delegate to Reader's scan logic (same implementation, avoids duplication)
      @reader._scan_offset(@reader.data, @offset, slot)
    end
  end
end

File: py/nxs.py
"""NXS Reader — zero-copy .nxb parser for Python.

Implements the Nexus Standard v1.0 binary wire format.

Usage:
    from nxs import NxsReader

    with open("data.nxb", "rb") as f:
        buf = f.read()                      # or mmap.mmap(...)
    reader = NxsReader(buf)

    reader.record_count                      # -> 1_000_000
    obj = reader.record(42)                  # O(1) jump
    obj.get_str("username")                  # decode one field

The reader does NOT materialize the full file. Each ``record(i)`` returns a
lightweight view; ``.get_*()`` decodes a single field on demand.
"""
from __future__ import annotations

import struct
from typing import Iterator, Optional, Union


# Magic bytes (little-endian u32)
MAGIC_FILE   = 0x4E585342  # NXSB
MAGIC_OBJ    = 0x4E58534F  # NXSO
MAGIC_LIST   = 0x4E58534C  # NXSL
MAGIC_FOOTER = 0x2153584E  # NXS!

# Sigil bytes
SIGIL_INT     = 0x3D  # =
SIGIL_FLOAT   = 0x7E  # ~
SIGIL_BOOL    = 0x3F  # ?
SIGIL_STR     = 0x22  # "
SIGIL_TIME    = 0x40  # @

# Pre-built struct unpackers (faster than struct.unpack_from with format string each call)
_U16 = struct.Struct("<H")
_U32 = struct.Struct("<I")
_U64 = struct.Struct("<Q")
_I64 = struct.Struct("<q")
_F64 = struct.Struct("<d")


class NxsError(Exception):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(f"{code}: {message}")
        self.code = code


class NxsReader:
    """Parses the preamble, schema, and tail-index of a .nxb buffer.

    The data sector is not walked — records are loaded lazily via ``record(i)``.
    """

    __slots__ = (
        "buf", "mv",
        "version", "flags", "dict_hash", "tail_ptr",
        "keys", "key_sigils", "key_index",
        "record_count", "_tail_start",
    )

    def __init__(self, buffer: Union[bytes, bytearray, memoryview]) -> None:
        if isinstance(buffer, memoryview):
            self.buf = buffer.tobytes() if not buffer.readonly else buffer
            self.mv = buffer
        else:
            self.buf = buffer
            self.mv = memoryview(buffer)

        if len(self.mv) < 32:
            raise NxsError("ERR_OUT_OF_BOUNDS", "file too small")

        # Preamble
        magic = _U32.unpack_from(self.mv, 0)[0]
        if magic != MAGIC_FILE:
            raise NxsError("ERR_BAD_MAGIC", f"expected NXSB, got 0x{magic:08x}")

        self.version   = _U16.unpack_from(self.mv, 4)[0]
        self.flags     = _U16.unpack_from(self.mv, 6)[0]
        self.dict_hash = _U64.unpack_from(self.mv, 8)[0]
        self.tail_ptr  = _U64.unpack_from(self.mv, 16)[0]

        # Footer check
        footer = _U32.unpack_from(self.mv, len(self.mv) - 4)[0]
        if footer != MAGIC_FOOTER:
            raise NxsError("ERR_BAD_MAGIC", "footer magic mismatch")

        # Schema (if embedded)
        self.keys: list[str] = []
        self.key_sigils: list[int] = []
        self.key_index: dict[str, int] = {}
        if self.flags & 0x0002:
            self._read_schema(32)

        # Tail-index
        self._read_tail_index()

    def _read_schema(self, offset: int) -> None:
        mv = self.mv
        key_count = _U16.unpack_from(mv, offset)[0]
        offset += 2
        self.key_sigils = list(mv[offset:offset + key_count])
        offset += key_count

        # Read null-terminated strings from the pool
        buf = bytes(mv[offset:])
        consumed = 0
        for _ in range(key_count):
            end = buf.index(0x00, consumed)
            self.keys.append(buf[consumed:end].decode("utf-8"))
            consumed = end + 1

        self.key_index = {k: i for i, k in enumerate(self.keys)}

    def _read_tail_index(self) -> None:
        p = self.tail_ptr
        self.record_count = _U32.unpack_from(self.mv, p)[0]
        self._tail_start = p + 4

    def record(self, i: int) -> "NxsObject":
        """O(1) lookup: get the top-level object at index ``i``."""
        if i < 0 or i >= self.record_count:
            raise NxsError("ERR_OUT_OF_BOUNDS",
                           f"record {i} out of [0, {self.record_count})")
        # Each entry: u16 keyId + u64 offset = 10 bytes
        entry = self._tail_start + i * 10
        abs_offset = _U64.unpack_from(self.mv, entry + 2)[0]
        return NxsObject(self, abs_offset)

    def records(self) -> Iterator["NxsObject"]:
        """Iterate all top-level records."""
        for i in range(self.record_count):
            yield self.record(i)


class NxsObject:
    """A lazy view over one NXS object. Fields are decoded on demand."""

    __slots__ = ("reader", "offset", "_parsed",
                 "_bitmask_bytes", "_offset_table_start", "length")

    def __init__(self, reader: NxsReader, offset: int) -> None:
        self.reader = reader
        self.offset = offset
        self._parsed = False

    def _parse_header(self) -> None:
        if self._parsed:
            return
        mv = self.reader.mv
        p = self.offset

        magic = _U32.unpack_from(mv, p)[0]
        if magic != MAGIC_OBJ:
            raise NxsError("ERR_BAD_MAGIC", f"expected NXSO at {p}")
        p += 4
        self.length = _U32.unpack_from(mv, p)[0]
        p += 4

        # LEB128 bitmask — read until high bit is 0
        bitmask: list[int] = []
        while True:
            b = mv[p]
            p += 1
            bitmask.append(b & 0x7F)
            if (b & 0x80) == 0:
                break

        self._bitmask_bytes = bitmask
        self._offset_table_start = p
        self._parsed = True

    def _field_offset(self, slot: int) -> Optional[int]:
        """Return the absolute byte offset of the field at ``slot``, or None."""
        self._parse_header()
        byte_idx, bit_idx = divmod(slot, 7)
        bitmask = self._bitmask_bytes
        if byte_idx >= len(bitmask):
            return None
        if not (bitmask[byte_idx] >> bit_idx) & 1:
            return None

        # Count present bits before this slot → position in offset table
        entry_idx = 0
        for s in range(slot):
            bi, bb = divmod(s, 7)
            if bi < len(bitmask) and (bitmask[bi] >> bb) & 1:
                entry_idx += 1

        rel = _U16.unpack_from(self.reader.mv,
                               self._offset_table_start + entry_idx * 2)[0]
        return self.offset + rel

    # ── Typed accessors ──────────────────────────────────────────────────────

    def get_i64(self, key: str) -> Optional[int]:
        slot = self.reader.key_index.get(key)
        if slot is None:
            return None
        off = self._field_offset(slot)
        if off is None:
            return None
        return _I64.unpack_from(self.reader.mv, off)[0]

    def get_f64(self, key: str) -> Optional[float]:
        slot = self.reader.key_index.get(key)
        if slot is None:
            return None
        off = self._field_offset(slot)
        if off is None:
            return None
        return _F64.unpack_from(self.reader.mv, off)[0]

    def get_bool(self, key: str) -> Optional[bool]:
        slot = self.reader.key_index.get(key)
        if slot is None:
            return None
        off = self._field_offset(slot)
        if off is None:
            return None
        return self.reader.mv[off] != 0

    def get_str(self, key: str) -> Optional[str]:
        slot = self.reader.key_index.get(key)
        if slot is None:
            return None
        off = self._field_offset(slot)
        if off is None:
            return None
        length = _U32.unpack_from(self.reader.mv, off)[0]
        return bytes(self.reader.mv[off + 4:off + 4 + length]).decode("utf-8")

    def get_time(self, key: str) -> Optional[int]:
        """Unix nanoseconds."""
        return self.get_i64(key)

    def to_dict(self) -> dict:
        """Decode all present fields (eager path)."""
        self._parse_header()
        out = {}
        for key, slot in self.reader.key_index.items():
            bi, bb = divmod(slot, 7)
            if bi >= len(self._bitmask_bytes):
                continue
            if not (self._bitmask_bytes[bi] >> bb) & 1:
                continue
            # Infer type from heuristic — in practice callers use typed accessors
            # This is best-effort for introspection.
            out[key] = self.get_str(key) if key in ("username", "email") else self.get_i64(key)
        return out
